### PR TITLE
HPCC-23461 Replace KeyIdentifier field with UserID field

### PR DESCRIPTION
**Pull request by mayx**
- Replace KeyIdentifier field with UserID field
- Require exact match on the UserID when trying to locate the key

Signed-off-by: Yanrui Ma <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mayx** [comment]
@shamser please review...

**Comment by richardkchapman** [comment]
Is mater (7.10.0) the correct target branch?

**Comment by mayx** [comment]
@richardkchapman yes 7.10.0 is good.

File: esp/services/ws_codesign/ws_codesignService.cpp [hunk 1]
```diff
@@ -47,17 +47,18 @@ bool Cws_codesignEx::onSign(IEspContext &context, IEspSignRequest &req, IEspSign
 {
     resp.setRetCode(-1);
 
-    StringBuffer keyid(req.getKeyIdentifier());
-    keyid.trim();
+    StringBuffer userid(req.getUserID());
```
Comment (shamser) [inline esp/services/ws_codesign/ws_codesignService.cpp]: The userid field needs to be XML decoded to make sure any characters are decoded correctly.

**Review by shamser — CHANGES_REQUESTED** [review]
@mayx

File: esp/services/ws_codesign/ws_codesignService.cpp [hunk 2]
```diff
@@ -47,17 +47,18 @@ bool Cws_codesignEx::onSign(IEspContext &context, IEspSignRequest &req, IEspSign
...
+    userid.trim();
     const char* text = req.getText();
-    if (keyid.length() == 0 || !text || !*text)
+    if (userid.length() == 0 || !text || !*text)
     {
         resp.setErrMsg("Please provide both KeyIdentifier and Text");
```
[the remark this answers is not in the document]
Comment (mayx) [reply esp/services/ws_codesign/ws_codesignService.cpp]: Fixed.

File: esp/services/ws_codesign/ws_codesignService.cpp (hunk 1 above)
Comment (mayx) [reply esp/services/ws_codesign/ws_codesignService.cpp:50]: ESP lower layer already did the decoding, so the value here is the xml-decoded value.

**Comment by mayx** [comment]
@shamser change made per suggestion, please review again...

**Review by shamser — APPROVED** [review]
@mayx Reviewed. Looks good.

**Pull request merged**